lib/batch_loader.rb

```diff
@@ -75,12 +77,20 @@ def __ensure_batched
     return if __executor_proxy.value_loaded?(item: @item)
 
     items = __executor_proxy.list_items
-    loader = ->(item, value) { __executor_proxy.load(item: item, value: value) }
+    loader =  -> (item, value = NULL_VALUE, &block) {
```

Comment:
It's not very common but there is another way to check whether the optional argument was passed or not (without checking the `value` itself):

```ruby
-> (item, value = (value_set = true ; nil), &block) {
  if block
    raise ... if value_set
    ...
  else
    raise ... unless value_set
    ...
  end
}
```

This way you can leave your second argument check (`raise`).

Reply:
That's neat, i didn't know that was possible in that context, although I don't find it particularly legible.

I also consider `(*item_and_maybe_value, &block)` followed by decomposition in the block, but didn't like that it changes the arity to be incredibly permissive.

I think the existing implementation is the most readable of the 3, if not necessarily the most clever.